src/GraphQLDataSource.ts

```diff
@@ -1,107 +1,67 @@
+      this.onRequestLink(),
+      this.link
+    ]);
   }
 
   public async mutation(mutation: DocumentNode, options: GraphQLRequest) {
     // GraphQL request requires the DocumentNode property to be named query
-    return this.executeSingleOperation({ ...options, query: mutation });
+    return this.execute({ ...options, query: mutation });
   }
 
   public async query(query: DocumentNode, options: GraphQLRequest) {
-    return this.executeSingleOperation({ ...options, query });
-  }
-
-  protected willSendRequest?(request: any): any;
-
-  private composeLinks(): ApolloLink {
-    const uri = this.resolveUri();
-
-    return ApolloLink.from([
-      this.onErrorLink(),
-      this.onRequestLink(),
-      createHttpLink({ fetch, uri }),
-    ]);
-  }
-
-  private didEncounterError(error: any) {
-    const status = error.statusCode ? error.statusCode : null;
-    const message = error.bodyText ? error.bodyText : null;
-
-    let apolloError: ApolloError;
-
-    switch (status) {
-      case 401:
-        apolloError = new AuthenticationError(message);
-        break;
-      case 403:
-        apolloError = new ForbiddenError(message);
-        break;
-      default:
-        apolloError = new ApolloError(message);
-    }
-
-    throw apolloError;
+    return this.execute({ ...options, query });
   }
 
-  private async executeSingleOperation(operation: GraphQLRequest) {
-    const link = this.composeLinks();
-
-    const [error, response] = await to(makePromise(execute(link, operation)));
-
-    if (error) {
-      this.didEncounterError(error);
-    }
-
-    return response;
-  }
-
-  private resolveUri(): string {
-    const baseURL = this.baseURL;
+  protected willSendRequest?(request: RequestOptions): ValueOrPromise<void>;
+
+  private async execute(operation: GraphQLRequest) {
+    try {
+      return await makePromise(execute(this.link, operation));
+    } catch (error) {
+      const status = error.statusCode ? error.statusCode : null;
+      const message = error.bodyText ? error.bodyText : null;
+
+      let apolloError: ApolloError;
+
+      switch (status) {
+        case 401:
+          apolloError = new AuthenticationError(message);
+          break;
+        case 403:
+          apolloError = new ForbiddenError(message);
+          break;
+        default:
+          apolloError = new ApolloError(message);
+      }
 
-    if (!baseURL) {
-      throw new ApolloError('Cannot make request to GraphQL API, missing baseURL');
+      throw apolloError;
     }
-
-    return baseURL;
   }
 
   private onRequestLink() {
-    return setContext((_, request) => {
+    // QUESTION: Is that first argument needed? Code for apollo mentions it's actually `request, prevContext`
+    return setContext(async (_, request: RequestOptions) => {
       if (this.willSendRequest) {
-        this.willSendRequest(request);
+        await this.willSendRequest(request);
       }
 
       return request;
     });
   }
-
-  private onErrorLink() {
```

Comment:
This should probably be handled by the user, as each user is like to have different error reporting requirements.